pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2020, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * {@link org.postgresql.core.Parser} stores information about placeholder occurrences in ParameterContext.
+ * In case of standard JDBC placeholders {@code '?'} the position in the SQL text is recorded. For named
+ * placeholders {@code ":paramName"} the name is recorded as well as the position. {@code PgPreparedStatement} can
+ * then use the name to lookup the parameter corresponding index. These values are also used by
+ * toString() methods to provide a human readable representation of the SQL text.
+ */
+public class ParameterContext {
+
+  enum BindStyle {
+    POSITIONAL, NAMED
+  }
+
+  public static final ParameterContext EMPTY_CONTEXT = new ParameterContext() {
```

Comment:
add javadoc to the constant describing that this is an immutable empty context which will throw exceptions on any attempt to add parameters.